arraycontext/impl/pytato.py

```diff
@@ -0,0 +1,536 @@
+"""
+.. currentmodule:: arraycontext
+.. autoclass:: PytatoArrayContext
+"""
+__copyright__ = """
+Copyright (C) 2020-1 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from arraycontext.fake_numpy import \
+        BaseFakeNumpyNamespace, BaseFakeNumpyLinalgNamespace
+from arraycontext.context import ArrayContext
+from arraycontext.container.traversal import \
+        rec_multimap_array_container, rec_map_array_container
+import numpy as np
+from typing import Any, Callable, Tuple, Union, Sequence, Mapping
+from pytools.tag import Tag
+import loopy as lp
+from dataclasses import dataclass, field
+from pyrsistent import pmap, PMap
+
+
+class _PytatoFakeNumpyLinalgNamespace(BaseFakeNumpyLinalgNamespace):
+    # Everything is implemented in the base class for now.
+    pass
+
+
+class _PytatoFakeNumpyNamespace(BaseFakeNumpyNamespace):
+    """
+    A :mod:`numpy` mimic for :class:`PytatoArrayContext`.
+
+
+    .. note::
+
+        :mod:`pytato` does not define any memory layout. If the caller invokes
+        any routine that depends on the input arrays' memory layout, a result is
+        returned assuming the arrays have a C-contiguous layout.
```

Comment:
I disagree with this. Anything that relies on memory layout should be an error, consistent with pytato: https://documen.tician.de/pytato/design.html#memory-layout.

Reply:
@inducer: The problem is other array-contexts rely on memory layouts, forcing pytato to have a memory layout. Downstream users also make use of the memory layout information: https://github.com/inducer/meshmode/blob/dab6ecffa61521ba3ab5fc72de2ad2a5fb431960/meshmode/dof_array.py#L670-L671. IMO, it cannot be avoided.

Reply:
I guess what I'm objecting to is the globality and specificity of this statement, which I find very restricting. AFAICT, this only applies to `ravel` right now, correct? My preferred solution would be to remove this statement and replace it with a pointer to the one in pytato, and to document `ravel(order="A")` as returning the array elements in "unspecified" order.

Reply:
Thanks for the clarification. I agree, we can just restrict this "unspecified layout" to ravel, rather than globally.

Reply:
0dc10bd